From e9c318d7dad2804c345afe83d6c1e0a31ea2d514 Mon Sep 17 00:00:00 2001
From: Lenka Doudova <lryznaro@redhat.com>
Date: Mon, 9 Jun 2025 15:15:04 +0200
Subject: [PATCH] Issue 6753 - Add 'add_exclude_subtree' and
 'remove_exclude_subtree' methods to Attribute uniqueness plugin

Description:
Adding 'add_exclude_subtree' and 'remove_exclude_subtree' methods to AttributeUniquenessPlugin in
order to be able to easily add or remove an exclude subtree.
Porting ticket 47927 test to
dirsrvtests/tests/suites/plugins/attruniq_test.py

Relates: #6753

Author: Lenka Doudova

Reviewers: Simon Pichugin, Mark Reynolds
---
 .../tests/suites/plugins/attruniq_test.py     |  16 ++
 dirsrvtests/tests/tickets/ticket47927_test.py | 267 ------------------
 src/lib389/lib389/plugins.py                  |  10 +
 3 files changed, 26 insertions(+), 267 deletions(-)
 delete mode 100644 dirsrvtests/tests/tickets/ticket47927_test.py

diff --git a/dirsrvtests/tests/suites/plugins/attruniq_test.py b/dirsrvtests/tests/suites/plugins/attruniq_test.py
index b338f405f..a2139c9d2 100644
--- a/dirsrvtests/tests/suites/plugins/attruniq_test.py
+++ b/dirsrvtests/tests/suites/plugins/attruniq_test.py
@@ -10,6 +10,7 @@ import pytest
 import ldap
 import logging
 from lib389.plugins import AttributeUniquenessPlugin
+from lib389.idm.nscontainer import nsContainers
 from lib389.idm.user import UserAccounts
 from lib389.idm.group import Groups
 from lib389._constants import DEFAULT_SUFFIX
@@ -21,6 +22,19 @@ logging.getLogger(__name__).setLevel(logging.DEBUG)
 log = logging.getLogger(__name__)
 MAIL_ATTR_VALUE = 'non-uniq@value.net'
 
+EXCLUDED_CONTAINER_CN = "excluded_container"
+EXCLUDED_CONTAINER_DN = "cn={},{}".format(EXCLUDED_CONTAINER_CN, DEFAULT_SUFFIX)
+
+EXCLUDED_BIS_CONTAINER_CN = "excluded_bis_container"
+EXCLUDED_BIS_CONTAINER_DN = "cn={},{}".format(EXCLUDED_BIS_CONTAINER_CN, DEFAULT_SUFFIX)
+
+ENFORCED_CONTAINER_CN = "enforced_container"
+
+USER_1_CN = "test_1"
+USER_2_CN = "test_2"
+USER_3_CN = "test_3"
+USER_4_CN = "test_4"
+
 
 def test_modrdn_attr_uniqueness(topology_st):
     """Test that we can not add two entries that have the same attr value that is
@@ -372,3 +386,5 @@ def test_matchingrule_attr(topology_st):
     log.info('Mod entry with exact case CN value which should be rejected')
     with pytest.raises(ldap.CONSTRAINT_VIOLATION):
         user.replace('cn', 'common_name')
+
+        attruniq.delete()
diff --git a/dirsrvtests/tests/tickets/ticket47927_test.py b/dirsrvtests/tests/tickets/ticket47927_test.py
deleted file mode 100644
index 887fe1af4..000000000
--- a/dirsrvtests/tests/tickets/ticket47927_test.py
+++ /dev/null
@@ -1,267 +0,0 @@
-# --- BEGIN COPYRIGHT BLOCK ---
-# Copyright (C) 2016 Red Hat, Inc.
-# All rights reserved.
-#
-# License: GPL (version 3 or any later version).
-# See LICENSE for details.
-# --- END COPYRIGHT BLOCK ---
-#
-import pytest
-from lib389.tasks import *
-from lib389.utils import *
-from lib389.topologies import topology_st
-
-from lib389._constants import SUFFIX, DEFAULT_SUFFIX, PLUGIN_ATTR_UNIQUENESS
-
-# Skip on older versions
-pytestmark = [pytest.mark.tier2,
-              pytest.mark.skipif(ds_is_older('1.3.4'), reason="Not implemented")]
-
-logging.getLogger(__name__).setLevel(logging.DEBUG)
-log = logging.getLogger(__name__)
-
-EXCLUDED_CONTAINER_CN = "excluded_container"
-EXCLUDED_CONTAINER_DN = "cn=%s,%s" % (EXCLUDED_CONTAINER_CN, SUFFIX)
-
-EXCLUDED_BIS_CONTAINER_CN = "excluded_bis_container"
-EXCLUDED_BIS_CONTAINER_DN = "cn=%s,%s" % (EXCLUDED_BIS_CONTAINER_CN, SUFFIX)
-
-ENFORCED_CONTAINER_CN = "enforced_container"
-ENFORCED_CONTAINER_DN = "cn=%s,%s" % (ENFORCED_CONTAINER_CN, SUFFIX)
-
-USER_1_CN = "test_1"
-USER_1_DN = "cn=%s,%s" % (USER_1_CN, ENFORCED_CONTAINER_DN)
-USER_2_CN = "test_2"
-USER_2_DN = "cn=%s,%s" % (USER_2_CN, ENFORCED_CONTAINER_DN)
-USER_3_CN = "test_3"
-USER_3_DN = "cn=%s,%s" % (USER_3_CN, EXCLUDED_CONTAINER_DN)
-USER_4_CN = "test_4"
-USER_4_DN = "cn=%s,%s" % (USER_4_CN, EXCLUDED_BIS_CONTAINER_DN)
-
-
-def test_ticket47927_init(topology_st):
-    topology_st.standalone.plugins.enable(name=PLUGIN_ATTR_UNIQUENESS)
-    try:
-        topology_st.standalone.modify_s('cn=' + PLUGIN_ATTR_UNIQUENESS + ',cn=plugins,cn=config',
-                                        [(ldap.MOD_REPLACE, 'uniqueness-attribute-name', b'telephonenumber'),
-                                         (ldap.MOD_REPLACE, 'uniqueness-subtrees', ensure_bytes(DEFAULT_SUFFIX)),
-                                         ])
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927: Failed to configure plugin for "telephonenumber": error ' + e.args[0]['desc'])
-        assert False
-    topology_st.standalone.restart(timeout=120)
-
-    topology_st.standalone.add_s(Entry((EXCLUDED_CONTAINER_DN, {'objectclass': "top nscontainer".split(),
-                                                                'cn': EXCLUDED_CONTAINER_CN})))
-    topology_st.standalone.add_s(Entry((EXCLUDED_BIS_CONTAINER_DN, {'objectclass': "top nscontainer".split(),
-                                                                    'cn': EXCLUDED_BIS_CONTAINER_CN})))
-    topology_st.standalone.add_s(Entry((ENFORCED_CONTAINER_DN, {'objectclass': "top nscontainer".split(),
-                                                                'cn': ENFORCED_CONTAINER_CN})))
-
-    # adding an entry on a stage with a different 'cn'
-    topology_st.standalone.add_s(Entry((USER_1_DN, {
-        'objectclass': "top person".split(),
-        'sn': USER_1_CN,
-        'cn': USER_1_CN})))
-    # adding an entry on a stage with a different 'cn'
-    topology_st.standalone.add_s(Entry((USER_2_DN, {
-        'objectclass': "top person".split(),
-        'sn': USER_2_CN,
-        'cn': USER_2_CN})))
-    topology_st.standalone.add_s(Entry((USER_3_DN, {
-        'objectclass': "top person".split(),
-        'sn': USER_3_CN,
-        'cn': USER_3_CN})))
-    topology_st.standalone.add_s(Entry((USER_4_DN, {
-        'objectclass': "top person".split(),
-        'sn': USER_4_CN,
-        'cn': USER_4_CN})))
-
-
-def test_ticket47927_one(topology_st):
-    '''
-    Check that uniqueness is enforce on all SUFFIX
-    '''
-    UNIQUE_VALUE = b'1234'
-    try:
-        topology_st.standalone.modify_s(USER_1_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_one: Failed to set the telephonenumber for %s: %s' % (USER_1_DN, e.args[0]['desc']))
-        assert False
-
-    # we expect to fail because user1 is in the scope of the plugin
-    try:
-        topology_st.standalone.modify_s(USER_2_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-        log.fatal('test_ticket47927_one: unexpected success to set the telephonenumber for %s' % (USER_2_DN))
-        assert False
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_one: Failed (expected) to set the telephonenumber for %s: %s' % (
-        USER_2_DN, e.args[0]['desc']))
-        pass
-
-    # we expect to fail because user1 is in the scope of the plugin
-    try:
-        topology_st.standalone.modify_s(USER_3_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-        log.fatal('test_ticket47927_one: unexpected success to set the telephonenumber for %s' % (USER_3_DN))
-        assert False
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_one: Failed (expected) to set the telephonenumber for %s: %s' % (
-        USER_3_DN, e.args[0]['desc']))
-        pass
-
-
-def test_ticket47927_two(topology_st):
-    '''
-    Exclude the EXCLUDED_CONTAINER_DN from the uniqueness plugin
-    '''
-    try:
-        topology_st.standalone.modify_s('cn=' + PLUGIN_ATTR_UNIQUENESS + ',cn=plugins,cn=config',
-                                        [(ldap.MOD_REPLACE, 'uniqueness-exclude-subtrees', ensure_bytes(EXCLUDED_CONTAINER_DN))])
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_two: Failed to configure plugin for to exclude %s: error %s' % (
-        EXCLUDED_CONTAINER_DN, e.args[0]['desc']))
-        assert False
-    topology_st.standalone.restart(timeout=120)
-
-
-def test_ticket47927_three(topology_st):
-    '''
-    Check that uniqueness is enforced on full SUFFIX except EXCLUDED_CONTAINER_DN
-    First case: it exists an entry (with the same attribute value) in the scope
-    of the plugin and we set the value in an entry that is in an excluded scope
-    '''
-    UNIQUE_VALUE = b'9876'
-    try:
-        topology_st.standalone.modify_s(USER_1_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_three: Failed to set the telephonenumber ' + e.args[0]['desc'])
-        assert False
-
-    # we should not be allowed to set this value (because user1 is in the scope)
-    try:
-        topology_st.standalone.modify_s(USER_2_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-        log.fatal('test_ticket47927_three: unexpected success to set the telephonenumber for %s' % (USER_2_DN))
-        assert False
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_three: Failed (expected) to set the telephonenumber for %s: %s' % (
-        USER_2_DN, e.args[0]['desc']))
-
-    # USER_3_DN is in EXCLUDED_CONTAINER_DN so update should be successful
-    try:
-        topology_st.standalone.modify_s(USER_3_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-        log.fatal('test_ticket47927_three: success to set the telephonenumber for %s' % (USER_3_DN))
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_three: Failed (unexpected) to set the telephonenumber for %s: %s' % (
-        USER_3_DN, e.args[0]['desc']))
-        assert False
-
-
-def test_ticket47927_four(topology_st):
-    '''
-    Check that uniqueness is enforced on full SUFFIX except EXCLUDED_CONTAINER_DN
-    Second case: it exists an entry (with the same attribute value) in an excluded scope
-    of the plugin and we set the value in an entry is in the scope
-    '''
-    UNIQUE_VALUE = b'1111'
-    # USER_3_DN is in EXCLUDED_CONTAINER_DN so update should be successful
-    try:
-        topology_st.standalone.modify_s(USER_3_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-        log.fatal('test_ticket47927_four: success to set the telephonenumber for %s' % USER_3_DN)
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_four: Failed (unexpected) to set the telephonenumber for %s: %s' % (
-        USER_3_DN, e.args[0]['desc']))
-        assert False
-
-    # we should be allowed to set this value (because user3 is excluded from scope)
-    try:
-        topology_st.standalone.modify_s(USER_1_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-    except ldap.LDAPError as e:
-        log.fatal(
-            'test_ticket47927_four: Failed to set the telephonenumber for %s: %s' % (USER_1_DN, e.args[0]['desc']))
-        assert False
-
-    # we should not be allowed to set this value (because user1 is in the scope)
-    try:
-        topology_st.standalone.modify_s(USER_2_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-        log.fatal('test_ticket47927_four: unexpected success to set the telephonenumber %s' % USER_2_DN)
-        assert False
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_four: Failed (expected) to set the telephonenumber for %s: %s' % (
-        USER_2_DN, e.args[0]['desc']))
-        pass
-
-
-def test_ticket47927_five(topology_st):
-    '''
-    Exclude the EXCLUDED_BIS_CONTAINER_DN from the uniqueness plugin
-    '''
-    try:
-        topology_st.standalone.modify_s('cn=' + PLUGIN_ATTR_UNIQUENESS + ',cn=plugins,cn=config',
-                                        [(ldap.MOD_ADD, 'uniqueness-exclude-subtrees', ensure_bytes(EXCLUDED_BIS_CONTAINER_DN))])
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_five: Failed to configure plugin for to exclude %s: error %s' % (
-        EXCLUDED_BIS_CONTAINER_DN, e.args[0]['desc']))
-        assert False
-    topology_st.standalone.restart(timeout=120)
-    topology_st.standalone.getEntry('cn=' + PLUGIN_ATTR_UNIQUENESS + ',cn=plugins,cn=config', ldap.SCOPE_BASE)
-
-
-def test_ticket47927_six(topology_st):
-    '''
-    Check that uniqueness is enforced on full SUFFIX except EXCLUDED_CONTAINER_DN
-    and EXCLUDED_BIS_CONTAINER_DN
-    First case: it exists an entry (with the same attribute value) in the scope
-    of the plugin and we set the value in an entry that is in an excluded scope
-    '''
-    UNIQUE_VALUE = b'222'
-    try:
-        topology_st.standalone.modify_s(USER_1_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_six: Failed to set the telephonenumber ' + e.args[0]['desc'])
-        assert False
-
-    # we should not be allowed to set this value (because user1 is in the scope)
-    try:
-        topology_st.standalone.modify_s(USER_2_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-        log.fatal('test_ticket47927_six: unexpected success to set the telephonenumber for %s' % (USER_2_DN))
-        assert False
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_six: Failed (expected) to set the telephonenumber for %s: %s' % (
-        USER_2_DN, e.args[0]['desc']))
-
-    # USER_3_DN is in EXCLUDED_CONTAINER_DN so update should be successful
-    try:
-        topology_st.standalone.modify_s(USER_3_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-        log.fatal('test_ticket47927_six: success to set the telephonenumber for %s' % (USER_3_DN))
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_six: Failed (unexpected) to set the telephonenumber for %s: %s' % (
-        USER_3_DN, e.args[0]['desc']))
-        assert False
-    # USER_4_DN is in EXCLUDED_CONTAINER_DN so update should be successful
-    try:
-        topology_st.standalone.modify_s(USER_4_DN,
-                                        [(ldap.MOD_REPLACE, 'telephonenumber', UNIQUE_VALUE)])
-        log.fatal('test_ticket47927_six: success to set the telephonenumber for %s' % (USER_4_DN))
-    except ldap.LDAPError as e:
-        log.fatal('test_ticket47927_six: Failed (unexpected) to set the telephonenumber for %s: %s' % (
-        USER_4_DN, e.args[0]['desc']))
-        assert False
-
-
-if __name__ == '__main__':
-    # Run isolated
-    # -s for DEBUG mode
-    CURRENT_FILE = os.path.realpath(__file__)
-    pytest.main("-s %s" % CURRENT_FILE)
diff --git a/src/lib389/lib389/plugins.py b/src/lib389/lib389/plugins.py
index 0878e18d1..3a7b26dae 100644
--- a/src/lib389/lib389/plugins.py
+++ b/src/lib389/lib389/plugins.py
@@ -175,6 +175,16 @@ class AttributeUniquenessPlugin(Plugin):
 
         self.set('uniqueness-across-all-subtrees', 'off')
 
+    def add_exclude_subtree(self, basedn):
+        """Add a uniqueness-exclude-subtrees attribute"""
+
+        self.add('uniqueness-exclude-subtrees', basedn)
+
+    def remove_exclude_subtree(self, basedn):
+        """Remove a uniqueness-exclude-subtrees attribute"""
+
+        self.remove('uniqueness-exclude-subtrees', basedn)
+
 
 class AttributeUniquenessPlugins(DSLdapObjects):
     """A DSLdapObjects entity which represents Attribute Uniqueness plugin instances
-- 
2.52.0

